### ISSUE TITLE
lkdemo.cpp crashing upon cornerSubPix() and calcOpticalFlowPyrLK() call

### ISSUE THREAD

**Issue by BrockTechnologies** [opencv/opencv#6648]
### Please state the information for your system
- OpenCV version: 3.0 and 3.1
- Host OS: Windows 7 Enterprise
- Microsoft Visual Studio 2008 and 2012
### In which part of the OpenCV library you got the issue?

Unknown.  Using the LKDemo.cpp File
### Expected behaviour

1.)Show Video, click and have green dot appear
2.) auto init have many dots appear
### Actual behaviour

two different errors when operating.  
lkdemo.cpp crashing upon cornerSubPix() and calcOpticalFlowPyrLK() call

I am able to build lkdemo.cpp in the samples folder of OpenCV 3.0. When I run the code, it displays the video from the webcam, but will not allow me to click and errors out with the following:

"Unhandled exception at 0x521ac3d2 in test.exe: 0xC0000005: Access violation reading location 0x00000000."

The code is stock. it stops at the following location: if( addRemovePt && points[1].size() < (size_t)MAX_COUNT ) { vector<point2f> tmp; tmp.push_back(point); -----> cornerSubPix( gray, tmp, winSize, Size(-1,-1), termcrit); points[1].push_back(tmp[0]); addRemovePt = false; }

Again stock code, I can see the video, but cannot click to put the dot on the screen. I also cannot auto-initialize.

I get the following when auto initializing: Unhandled exception at 0x5552df22 (msvcr110.dll) in test.exe: 0xC0000005: Access violation writing location 0x00000000.
### Code

``` .cpp
#include "opencv2/video/tracking.hpp"
#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/highgui/highgui.hpp"

#include <iostream>
#include <ctype.h>

using namespace cv;
using namespace std;

static void help()
{
    // print a welcome message, and the OpenCV version
    cout << "\nThis is a demo of Lukas-Kanade optical flow lkdemo(),\n"
            "Using OpenCV version " << CV_VERSION << endl;
    cout << "\nIt uses camera by default, but you can provide a path to video as an argument.\n";
    cout << "\nHot keys: \n"
            "\tESC - quit the program\n"
            "\tr - auto-initialize tracking\n"
            "\tc - delete all the points\n"
            "\tn - switch the \"night\" mode on/off\n"
            "To add/remove a feature point click it\n" << endl;
}

Point2f point;
bool addRemovePt = false;

static void onMouse( int event, int x, int y, int /*flags*/, void* /*param*/ )
{
    if( event == CV_EVENT_LBUTTONDOWN )
    {
        point = Point2f((float)x, (float)y);
        addRemovePt = true;
    }
}

int main( int argc, char** argv )
{
    help();

    VideoCapture cap;
    TermCriteria termcrit(CV_TERMCRIT_ITER|CV_TERMCRIT_EPS, 20, 0.03);
    Size subPixWinSize(10,10), winSize(31,31); // <------------------------------ subPixWinSize

    const int MAX_COUNT = 500;
    bool needToInit = false;
    bool nightMode = false;

    if( argc == 1 || (argc == 2 && strlen(argv[1]) == 1 && isdigit(argv[1][0])))
        cap.open(argc == 2 ? argv[1][0] - '0' : 0);
    else if( argc == 2 )
        cap.open(argv[1]);

    if( !cap.isOpened() )
    {
        cout << "Could not initialize capturing...\n";
        return 0;
    }

    namedWindow( "LK Demo", 1 );
    setMouseCallback( "LK Demo", onMouse, 0 );

    Mat gray, prevGray, image;
    vector<point2f> points[2];

    for(;;) // <------------------------- THE FOR LOOP
    {
        Mat frame;
        cap >> frame;
        if( frame.empty() )
            break;

        frame.copyTo(image);
        cvtColor(image, gray, COLOR_BGR2GRAY);

        if( nightMode )
            image = Scalar::all(0);

        if( needToInit )
        {
            // automatic initialization
            goodFeaturesToTrack(gray, points[1], MAX_COUNT, 0.01, 10, Mat(), 3, 0, 0.04);
            cornerSubPix(gray, points[1], subPixWinSize, Size(-1,-1), termcrit);
            addRemovePt = false;
        }
        else if( !points[0].empty() )
        {
            vector<uchar> status;
            vector<float> err;
            if(prevGray.empty())
                gray.copyTo(prevGray);
// ==================================== CRASH B ================================================
            calcOpticalFlowPyrLK(prevGray, gray, points[0], points[1], status, err, winSize,
                                 3, termcrit, 0, 0.001);
            size_t i, k;
            for( i = k = 0; i < points[1].size(); i++ )
            {
                if( addRemovePt )
                {
                    if( norm(point - points[1][i]) <= 5 )
                    {
                        addRemovePt = false;
                        continue;
                    }
                }

                if( !status[i] )
                    continue;

                points[1][k++] = points[1][i];
                circle( image, points[1][i], 3, Scalar(0,255,0), -1, 8);
            }
            points[1].resize(k);
        }

        if( addRemovePt && points[1].size() < (size_t)MAX_COUNT )
        {
            vector<point2f> tmp;
            tmp.push_back(point);
// ===================================== CRASH A =================================================
            cornerSubPix( gray, tmp, winSize, cvSize(-1,-1), termcrit);
            points[1].push_back(tmp[0]);
            addRemovePt = false;
        }

        needToInit = false;
        imshow("LK Demo", image); // <--------------------------- imshow()

        char c = (char)waitKey(10);
        if( c == 27 )
            break;
        switch( c )
        {
        case 'r':
            needToInit = true;
            break;
        case 'c':
            points[0].clear();
            points[1].clear();
            break;
        case 'n':
            nightMode = !nightMode;
            break;
        }

        std::swap(points[1], points[0]);


        cv::swap(prevGray, gray);
    }

    return 0;
}
```

I have tried the waitKey(1000); solution  with no luck

**Comment by NCBee** [comment]
I get access violation on the next [seemingly unrelated] operation after calling cornerSubPix() successfully in my app. Commenting out the cornerSubPix fixes the problem so there is definitely something fishy happening in there.

**Comment by souch55** [comment]
Hi NCBee can you give more detail about the error you are facing and source code of the app where you are getting access violation error after calling cornerSubPix().

**Comment by BrockTechnologies** [comment]
What additional information is needed?

R,

Keith M. Brock
520.271.7566 Cell


On Jul 22, 2016, at 02:11, souch55 <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi NCBee can you give more detail about the error you are facing and source code of the app where you are getting access violation error after calling cornerSubPix().

-

**Comment by mshabunin** [comment]
Can not reproduce with latest OpenCV version on Ubuntu 16.04 (GCC) and Win 10 (VS 2015), `lkdemo` sample app works fine, without any crashes.

**Comment by alalek** [comment]
Probably crash was fixed here: #7206

**Comment by BrockTechnologies** [comment]
Please read what revision and os we were using.

R,

Keith M. Brock
520.271.7566 Cell


On Dec 5, 2016, at 07:38, Alexander Alekhin <notifications@github.com<mailto:notifications@github.com>> wrote:


Probably crash was fixed here: #7206<https://github.com/opencv/opencv/pull/7206>

-

**Comment by alalek** [comment]
@BrockTechnologies Did mentioned patch resolves your problem?
